subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AbstractTransformer.java

```diff
@@ -18,27 +18,96 @@
+import org.gradle.api.Project;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.api.internal.project.ProjectStateRegistry;
+import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
+import org.gradle.api.internal.tasks.WorkNodeAction;
+import org.gradle.internal.classloader.ClassLoaderHierarchyHasher;
 import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.isolation.Isolatable;
 
+import javax.annotation.Nullable;
 import java.io.File;
 
-public abstract class AbstractTransformer<T> implements Transformer {
+public abstract class AbstractTransformer<T, P> implements Transformer {
     private final Class<? extends T> implementationClass;
-    private final HashCode inputsHash;
     private final ImmutableAttributes fromAttributes;
+    private final DomainObjectContextProjectStateHandler projectStateHandler;
+    private IsolatableParameters<P> isolatableParameters;
+    private final ProjectStateRegistry.SafeExclusiveLock isolationLock;
+    private WorkNodeAction isolateAction;
 
-    public AbstractTransformer(Class<? extends T> implementationClass, HashCode inputsHash, ImmutableAttributes fromAttributes) {
+    public AbstractTransformer(Class<? extends T> implementationClass, ImmutableAttributes fromAttributes, DomainObjectContextProjectStateHandler projectStateHandler) {
         this.implementationClass = implementationClass;
-        this.inputsHash = inputsHash;
         this.fromAttributes = fromAttributes;
+        this.projectStateHandler = projectStateHandler;
+        this.isolationLock = projectStateHandler.newExclusiveOperationLock();
+        this.isolateAction = new WorkNodeAction() {
+            @Nullable
+            @Override
+            public Project getProject() {
+                return projectStateHandler.maybeGetOwningProject();
+            }
+
+            @Override
+            public void run() {
+                isolateExclusively();
+            }
+        };
+    }
+
+    static class IsolatableParameters<T> {
+        private HashCode secondaryInputsHash;
+        private Isolatable<T> isolatableParameters;
+
+        public IsolatableParameters(Isolatable<T> isolatableParameters, HashCode secondaryInputsHash) {
+            this.secondaryInputsHash = secondaryInputsHash;
+            this.isolatableParameters = isolatableParameters;
+        }
+
+        public HashCode getSecondaryInputsHash() {
+            return secondaryInputsHash;
+        }
+
+        public Isolatable<T> getIsolatableParameters() {
+            return isolatableParameters;
+        }
     }
 
     @Override
     public ImmutableAttributes getFromAttributes() {
         return fromAttributes;
     }
 
+    abstract protected IsolatableParameters<P> doIsolateParameters();
+
+    @Override
+    public void visitDependencies(TaskDependencyResolveContext context) {
+        context.add(isolateAction);
```

Comment:
It would be good (maybe in a later PR) to not schedule this node if it won't do anything

Reply:
The node is now only scheduled if it actually has parameters which need to be isolated.